crates/hir-def/src/body/lower.rs

```diff
@@ -100,6 +101,8 @@ struct ExprCollector<'a> {
     // resolution
     label_ribs: Vec<LabelRib>,
     current_binding_owner: Option<ExprId>,
+
+    awaitable_blocks: Vec<Awaitable>,
```

Comment:
Hmm, an `Option<Awaitable>` should suffice here, which would be replaced with `coroutine_kind` after https://github.com/rust-lang/rust-analyzer/pull/16173. In fact you can probably take a peek at how that PR sets `coroutine_kind`. Mihgt be simple enough to copy if the changes aren't drastic in there.

Then we just need to swap whenever we enter and exit a context like we do with other `Option`/`bool` things here